### ISSUE TITLE
Declare core.iterators package public API

### ISSUE THREAD

**Issue by milleruntime** [apache/accumulo#1390]
This is something that is long overdue but has been put off for good reasons.  Some refactoring was done in 2.0 for classes in this package to not use internal types.  SortedKeyValueIterator should be declared public API and I think this package is close to this goal where it could be done for 2.1

**Comment by milleruntime** [comment]
Classes under core.iterators with *impl* imports:
```
LongCombiner.java:import org.apache.accumulo.core.clientImpl.lexicoder.AbstractLexicoder;
system/MapFileIterator.java:import org.apache.accumulo.core.sample.impl.SamplerConfigurationImpl;
system/SequenceFileIterator.java:import org.apache.accumulo.core.sample.impl.SamplerConfigurationImpl;
system/MultiIterator.java:import org.apache.accumulo.core.dataImpl.KeyExtent;
user/TimestampFilter.java:import java.text.SimpleDateFormat;
user/BigDecimalCombiner.java:import org.apache.accumulo.core.clientImpl.lexicoder.AbstractLexicoder;
user/SummingArrayCombiner.java:import org.apache.accumulo.core.clientImpl.lexicoder.AbstractLexicoder;
```

**Comment by ctubbsii** [comment]
It's okay to have *impl* imports. It's just not okay for them to expose those types in their public/protected members. apilyzer-maven-plugin should catch violations of this.

**Comment by ctubbsii** [comment]
We should probably move system iterators out of the way, because those should not be public API, but otherwise, I think this is fine.

**Comment by milleruntime** [comment]
Based on some discussion in #1415, iterators should be treated as SPI instead.

**Comment by milleruntime** [comment]
This is dependent on #1426

**Comment by keith-turner** [comment]
I made a few changes to #1426 and moved it from draft to ready for review.

**Comment by milleruntime** [comment]
I think the work that needed to be done in the code for this is done.  So closing.  There is a separate ticket for documentation that is still open #1416
